packages/common/client-utils/src/layerCompat.ts

```diff
@@ -0,0 +1,144 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Result of a layer compatibility check - whether a layer is compatible with another layer.
+ * @internal
+ */
+export type LayerCompatCheckResult =
+	| { readonly isCompatible: true }
+	| {
+			readonly isCompatible: false;
+			/**
+			 * Whether the generation of the layer is compatible with the other layer.
+			 */
+			readonly isGenerationCompatible: boolean;
+			/**
+			 * The features that are required by the layer but are not supported by the other layer. This will only
+			 * be set if there are unsupported features.
+			 */
+			readonly unsupportedFeatures: string[] | undefined;
+	  };
+
+/**
+ * @internal
+ */
+export const ICompatibilityDetails: keyof IProvideCompatibilityDetails =
+	"ICompatibilityDetails";
+
+/**
+ * @internal
+ */
+export interface IProvideCompatibilityDetails {
+	readonly ICompatibilityDetails: ICompatibilityDetails;
+}
+
+/**
+ * This interface is used to communicate the compatibility details of a layer to another layer.
+ * @internal
+ */
+export interface ICompatibilityDetails extends IProvideCompatibilityDetails {
```

Comment:
might want to consider making it `extends Partial<IProvideCompatibilityDetails>` then not all compat details need to be providers, which can make it a bit easier to work with the types

Reply:
Great suggestion. This made working with it a whole lot easier.